File: pset6/credit/credit.py
from cs50 import get_int
number = -1
while number < 0:
    number = get_int("Number: ")
# Luhn's algorithm
# checking if the credit card is valid
sum = number % 10
# determining the length of the number
s_num = str(number)
length = len(s_num)
for i in range(length):
    # determine all the numbers
    m = 10**(i+2)
    l = 10**(i+1)
    num = (number % m - number % l) / l
    # multiply every second by 2
    if i == 0 or i % 2 == 0:
        if (num > 4):
            num = (num * 2) % 10 + 1
        elif num <= 4:
            num = num * 2
    sum += num

# determing first numbers
start_number = number
for i in range(length - 2):
    start_number = int(start_number/10)
# checking the validity of the card
if sum % 10 == 0:
    if (length == 15) and (start_number == 34 or (start_number == 37)):
        print("AMEX")
    elif (length == 16) and (start_number >= 51 and start_number <= 55):
        print("MASTERCARD")
    elif (length == 13 or length == 16) and (start_number >= 40 and start_number < 50):
        print("VISA")
    else:
        print("INVALID")
else:
    print("INVALID")

File: pset1/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>
int main(void)
{
    //Ask for card number
    long number;
    do
    {
        number = get_long("Number:");
    }
    while (number < 0);
    //Luhn's algorithm
    //checking of the credit card is valid
    int sum = number % 10;
    for (int i = 0; i < 15; i++)
    {
        //determine all the numbers
        long m = pow(10, i + 2);
        long l = pow(10, i + 1);
        int num = (number % m - number % l) / l;
        //multiply every second by 2
        if ((i == 0) || (i == 2) || (i == 4) || (i == 6) || (i == 8) || (i == 10) || (i == 12) || (i == 14))
        {
            if (num > 4)
            {
                num = (num * 2) % 10 + 1;
            }
            else if (num <= 4)
            {
                num = num * 2;
            }
        }
        sum = sum + num;
    }
    //determining the lenght od the number
    int length = 0;
    for (long i = 1; i < number; i *= 10)
    {
        length++;
    }
    // determining first numbers
    long start_number = number;
    for (int i = 0; i < (length - 2); i++)
    {
        start_number = start_number / 10;
    }
    //checking the validity of the card
    if (sum % 10 == 0)
    {
        if ((length == 15) && ((start_number == 34) || (start_number == 37)))
        {
            printf("AMEX\n");
        }
        else if ((length == 16) && (start_number >= 51) && (start_number <= 55))
        {
            printf("MASTERCARD\n");
        }
        else if (((length == 13) || (length == 16)) && ((start_number >= 40) && (start_number < 50)))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}